### PR TITLE
Bug 1821005 - Upgrade Gradle to 8.3 version and Android Gradle Plugin to 8.0.2

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1821005
https://bugzilla.mozilla.org/show_bug.cgi?id=1849833

### CODE REVIEW

File: android-components/components/support/base/build.gradle
Comment (gbrownmozilla) [inline android-components/components/support/base/build.gradle:8]: Just out of curiosity: Wouldn't `layout.buildDirectory` work just as well here?

File: fenix/app/src/main/AndroidManifest.xml
Comment (gbrownmozilla) [inline fenix/app/src/main/AndroidManifest.xml:46]: I see this is now useLegacyPackaging, https://developer.android.com/guide/topics/manifest/application-element#extractNativeLibs - lgtm.

**Review by gbrownmozilla — APPROVED** [review]
I don't know much about Gradle, but this works for me and looks just fine. Thanks!

File: android-components/components/support/base/build.gradle
Comment (mcarare) [reply android-components/components/support/base/build.gradle:8]: That would throw a runtime exception as` buildDirectory` is a `DirectoryProperty` and there is no `File` constructor that takes that combination of arguments.

File: android-components/samples/browser/src/main/AndroidManifest.xml
Comment (jonalmeida) [inline android-components/samples/browser/src/main/AndroidManifest.xml:8]: Leaving a comment here, but it applies to all the other apps in this commit:

So I think this is not enough for us.

We use the camera permission for the front/back/external cameras.

What I'm reading on [the docs here](https://developer.android.com/guide/topics/manifest/uses-feature-element#hw-features), is that this permission is only for the back camera.

We either need to use `android.hardware.camera.any` instead, or we need to add three separate `uses-feature` for `android.hardware.camera`, `android.hardware.camera.front`, and `android.hardware.camera.external`. It's not clear to me from this doc at least which one we need.

Comment (jonalmeida) [inline android-components/samples/browser/src/main/AndroidManifest.xml:8]: Should this be in the library module that users these hardware features instead?

File: fenix/gradle.properties
Comment (jonalmeida) [inline fenix/gradle.properties:31]: > From Gradle 8 the default value has changed from false to true.

Does this mean that we are failing to use non-transitive R classes and non-final res Ids? Can we file a follow-up to fix this if that's true?

**Review by jonalmeida — COMMENTED** [review]
Thanks for separate commits and comments, it made this review very easy and less intimidating!

Left some open questions, but once addressed, we could land this.

File: fenix/build.gradle
Comment (mcarare) [reply fenix/build.gradle:180]: Because otherwise, building projects that do not have the `android` property would throw a `MissingMethodException`, as the extension does not exist in those types of Gradle projects.

File: fenix/gradle.properties
Comment (mcarare) [reply fenix/gradle.properties:31]: Yes, we don't use them in Fenix or Focus. Moving to non-transitive R classes and non-final res Ids implies changing hundreds of files, so I thought it would be best not to do it here. And we still should decide if we want to.

File: android-components/samples/browser/src/main/AndroidManifest.xml
Comment (mcarare) [reply android-components/samples/browser/src/main/AndroidManifest.xml:8]: This was added because the manifest specifies a permission that [implies](https://developer.android.com/topic/arc/manifest#implied-features) a feature requirment. Meaning that without specifying that the feature is not required, it would imply that it is required, and the app would not be able to be found on devices without that specific hardware (in our case Chromebook).
We could also suppress the lint warning but this seems more clean.

Comment (mcarare) [reply android-components/samples/browser/src/main/AndroidManifest.xml:8]: I originally thought so myself, but I preferred to add it to the app manifests as the permission was declared directly in the manifest, not in the merged one; so for visibility sake.

File: fenix/gradle.properties
Comment (jonalmeida) [reply fenix/gradle.properties:31]: Sounds good, let's file a bug then so we can track if we want to do it or not. I'm not sure what either direction implies right now.

File: android-components/samples/browser/src/main/AndroidManifest.xml
Comment (jonalmeida) [reply android-components/samples/browser/src/main/AndroidManifest.xml:8]: > Meaning that without specifying that the feature is not required, it would imply that it is required, and the app would not be able to be found on devices without that specific hardware (in our case Chromebook).

Even then, we would need to say that we need (non-required) hardware to all cameras in web content, right?

File: fenix/gradle.properties
Comment (mcarare) [reply fenix/gradle.properties:31]: Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1863329 and https://bugzilla.mozilla.org/show_bug.cgi?id=1863330

File: android-components/samples/browser/src/main/AndroidManifest.xml
Comment (mcarare) [reply android-components/samples/browser/src/main/AndroidManifest.xml:12]: Added a comment and commit message to make this more clear for future references.

**Review by jonalmeida — APPROVED** [review]

**Pull request merged**
